File: Algorithms/0890_Find_and_Replace_Pattern/Python/Find_and_Replace_Pattern_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def findAndReplacePattern(self, words, pattern):
        cache = {}
        res = []
        for word in words:
            if len(word) != len(pattern): continue
            temp_cache = {}
            for i in range(len(word)):
                if word[i] not in temp_cache:
                    cache[pattern[i]] = word[i]
                    temp_cache[word[i]] = pattern[i]
                elif word[i] in temp_cache and temp_cache[word[i]] != pattern[i]:
                    break
            else:
                temp = ''.join(cache[c] for c in pattern)
                if temp == word: res.append(word)

        return res

File: Algorithms/0890_Find_and_Replace_Pattern/Java/FindandReplacePatternSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> res = new ArrayList<>();
        Map<Character, Character> cache = new HashMap<>();
        for (String word : words) {
            if (word.length() != pattern.length()) continue;
            Map<Character, Character> temp = new HashMap<>();
            boolean flag = true;
            for (int i = 0; i < word.length(); i++) {
                if (temp.containsKey(word.charAt(i)) && temp.get(word.charAt(i)) != pattern.charAt(i)) {
                    flag = false;
                    break;
                }
                if (!temp.containsKey(word.charAt(i))) {
                    cache.put(pattern.charAt(i), word.charAt(i));
                    temp.put(word.charAt(i), pattern.charAt(i));
                }
            }
            if (flag) {
                StringBuilder tempRes = new StringBuilder();
                for (char c : pattern.toCharArray()) {
                    tempRes.append(cache.get(c));
                }
                if (word.equals(tempRes.toString())) res.add(word);
            }
        }
        return res;
    }
}





